### ISSUE TITLE
Feature request: Microphone audio transcribed into active window.

### ISSUE THREAD

**Issue by glinkot** [Const-me/Whisper#157]
The performance of this project is fantastic!  What would be a great use of this is as the title says - you hit a key, speak your phrase, and it enters the spoken text it at the current cursor location.  

This exact idea has been implemented here (https://github.com/savbell/whisper-writer) but the latency makes that one challenging to make use of.

**Comment by tigros** [comment]
Hi,

I decided to make one for you just for fun.

It borrows a couple ideas from whisperer, let me know how it goes.

https://github.com/tigros/livewriter

**Comment by glinkot** [comment]
Wow, fantastic stuff tigros, I can't believe you'd do that!  So awesome that people like you are in the community.  Luckily my least bad language is C#, I'm not great at it but I'll try to learn from what it does.  Could I ask a couple of questions?

- is main.exe the command line whisper.exe?

- it seems to wait the specified seconds, then process the captured wav by executing the command line. I imagine it has to load the model file once for each round.  Do you think it's plausible to use the whispernet package, so that it 'warms up' (loads model into memory etc and establishes context) so it is quicker each time?  Not asking you to do it, just seeing if you think it's workable.

- Do you know any ways to determine silence using those NAudio libraries?  The intended use case here is one shot- say you're using slack or whatsapp, whatever, you put the cursor where you want, hit the hotkey, say a brief thing and once a period of silence has happened it recognises/types it into the box with minimal waiting once the sentence is finished.  This version runs more as a toggle with batching of wav capture, and types [BLANK_AUDIO] or [Silence] when it doesn't hear anything.  

For reference, the one I referred to earlier seems to capture 'frames' of audio (between 10 and 30ms) and detects silence as a sequence of >n silent frames (https://github.com/savbell/whisper-writer/blob/main/src/transcription.py).  An interesting way to do it.

Trying to make some of these adjustments could be an interesting exercise for me, now you've done the difficult part!  

Thanks again tigros, so cool what you've created.

Mark

**Comment by tigros** [comment]
Hi Mark,

You're welcome, it was fun!

I don't know how to load the model once and use it for live audio, in C#, but I'm sure it's doable in C++, fun for someone else..

I tried to detect speaking/silence at first, didn't work well when they don't shut up like a radio program I used to test.

After using this GPU version I wouldn't want to go back to CPU, too much faster, I'll look at whisper.net looks interesting but surely not as fast as this Const-me version.

Not very familiar with NAudio, this was first time using it, to detect silence I was just adding up the byte values looking for a threshold. It's doing that now to skip running main.exe if quiet.

I don't mind making small improvements here and there, but loading model once in C++ etc is a bit much, love to see Const-me do that myself.

**Comment by glinkot** [comment]
The whisper-net I meant is a folder within the Const-Me project (https://github.com/Const-me/Whisper/tree/master/WhisperNet), that gives you the .NET native interop, so I imagine performance would be comparable through that wrapper.  There's a nuget package published by ConstMe (https://www.nuget.org/packages/WhisperNet) I thought might be good.

**Comment by tigros** [comment]
ah yes i see now a Transcribe.cs and related stuff, so there's hope! I just went with an idea having whisperer and main.exe in mind. I'll have to take a closer look at all that.

**Comment by glinkot** [comment]
Yes, it does look a bit hopeful.  I've been looking around today and found this also, the voice activity detection used by the other project was a python port of a google project, and there is a .net one also (https://github.com/ladenedge/WebRtcVadSharp).  So between that and the transcribe thing, I think it's plausible!  Easy, perhaps not so much, but doable!

**Comment by emcodem** [comment]
Just a sidenote, VAD is already builtin, it is utilized also in the existing MicrophoneCS example...

**Comment by glinkot** [comment]
Hi there,

I decided to spend a bit of time working this into a UI.  In the end I found the MicrophoneCS example and the library a bit hard to get my head around as a more basic coder, so I ended up using Whisper.net for the actual transcription.  I'd like to support both though, and would like to add continuous transcription at some point too.

It's on github [here](https://github.com/glinkot/WhisperType)
